test/Microsoft.ML.Sweeper.Tests/TestSweeper.cs

```diff
@@ -0,0 +1,664 @@
+                    task.Wait();
+                    Assert.Equal(TaskStatus.RanToCompletion, task.Status);
+                    var paramWithId = task.Result;
+                    if (paramWithId == null)
+                        return;
+                    Thread.Sleep(sleeps[i]);
+                    var result = new RunResult(paramWithId.ParameterSet, 0.42, true);
+                    sweeper.Update(paramWithId.Id, result);
+                    lock (mlock)
+                        paramSets.Add(paramWithId.ParameterSet);
+                });
+                Assert.True(paramSets.Count <= sweeps);
+                CheckAsyncSweeperResult(paramSets);
+            }
+        }
+
+        [Fact]
+        public void TestNelderMeadSweeperAsync()
+        {
+            var random = new Random(42);
+            using (var env = new TlcEnvironment(42))
+            {
+                int batchSize = 5;
+                int sweeps = 40;
+                var paramSets = new List<ParameterSet>();
+                var args = new DeterministicSweeperAsync.Arguments();
+                args.BatchSize = batchSize;
+                args.Relaxation = 0;
+
+                args.Sweeper = ComponentFactoryUtils.CreateFromFunction(
+                    environ => {
+                        var param = new IComponentFactory<INumericValueGenerator>[] {
+                            ComponentFactoryUtils.CreateFromFunction(
+                                innerEnviron => new FloatValueGenerator(new FloatParamArguments() { Name = "foo", Min = 1, Max = 5})),
+                            ComponentFactoryUtils.CreateFromFunction(
+                                innerEnviron => new LongValueGenerator(new LongParamArguments() { Name = "bar", Min = 1, Max = 1000, LogBase = true }))
+                        };
+
+                        var nelderMeadSweeperArgs = new NelderMeadSweeper.Arguments()
+                        {
+                            SweptParameters = param,
+                            FirstBatchSweeper = ComponentFactoryUtils.CreateFromFunction<IValueGenerator[], ISweeper>(
+                                (firstBatchSweeperEnviron, firstBatchSweeperArgs) =>
+                                    new RandomGridSweeper(environ, new RandomGridSweeper.Arguments() { SweptParameters = param }))                
+                        };
+
+                        return new NelderMeadSweeper(environ, nelderMeadSweeperArgs);
+                    }
+                );
+           
+                var sweeper = new DeterministicSweeperAsync(env, args);
+                var mlock = new object();
+                double[] metrics = new double[sweeps];
+                for (int i = 0; i < metrics.Length; i++)
+                    metrics[i] = random.NextDouble();
+
+                for (int i = 0; i < sweeps; i++)
+                {
+                    var task = sweeper.Propose();
+                    task.Wait();
```

Comment:
How about using `public async Task TestNelderMeadSweeperAsync()` in the method signature, and then you can just `await sweeper.Propose()` here?